compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp

```diff
@@ -0,0 +1,108 @@
+  auto producer = v.getDefiningOp();
+  while (auto castProducer = dyn_cast<tensor::CastOp>(producer)) {
+    producer = castProducer.getSource().getDefiningOp();
+  }
+  return producer;
+}
+
+void ConfigTrackingListener::notifyOperationReplaced(Operation *op,
+                                                     ValueRange replacement) {
+  // We have no way to track replacements without a producer.
+  if (replacement.empty()) {
+    return;
+  }
+
+  IREE::Codegen::LoweringConfigAttrInterface loweringConfig =
+      getLoweringConfig(op);
+  if (!loweringConfig) {
+    return;
+  }
+
+  // Must have a producer of the same type to track the lowering config.
+  auto producer = skipCastsDefiningOp(replacement.front());
+  if (!producer || producer->getName() != op->getName()) {
+    return;
+  }
+
+  for (auto v : replacement.drop_front()) {
+    // Conservatively require that all replacements are produced by the same
+    // operation.
+    if (skipCastsDefiningOp(v) != producer) {
+      return;
+    }
+  }
+
+  // No need to add the lowering config if it's already present.
+  if (getLoweringConfig(producer)) {
+    return;
+  }
+
+  setLoweringConfig(producer, loweringConfig);
+}
+
+namespace {
+
+/// Add the corresponding fast-math flags to operations given a floating-point
+/// optimization mode.
+// TODO: For now we only allow default flags, such as arithmetic reassociation.
+struct ConfigTrackingCanonicalizerPass final
+    : impl::ConfigTrackingCanonicalizerPassBase<
+          ConfigTrackingCanonicalizerPass> {
+public:
+  using impl::ConfigTrackingCanonicalizerPassBase<
+      ConfigTrackingCanonicalizerPass>::ConfigTrackingCanonicalizerPassBase;
+  /// Initialize the canonicalizer by building the set of patterns used during
+  /// execution.
+  LogicalResult initialize(MLIRContext *context) override {
+    // Inherit the same config defaults from the upstream canonicalizer pass.
+    config.useTopDownTraversal = true;
+    config.enableRegionSimplification = mlir::GreedySimplifyRegionLevel::Normal;
+
+    RewritePatternSet owningPatterns(context);
+    for (auto *dialect : context->getLoadedDialects())
+      dialect->getCanonicalizationPatterns(owningPatterns);
+    for (RegisteredOperationName op : context->getRegisteredOperations())
+      op.getCanonicalizationPatterns(owningPatterns, context);
+
+    patterns =
+        std::make_shared<FrozenRewritePatternSet>(std::move(owningPatterns));
+    return success();
+  }
+
+  void runOnOperation() override {
+    // Canonicalization is best-effort. Non-convergence is not a pass failure.
+    auto listener = ConfigTrackingListener();
+    config.listener = &listener;
+    LogicalResult didConverge =
+        applyPatternsAndFoldGreedily(getOperation(), *patterns, config);
+    if (this->testConvergence && failed(didConverge)) {
+      getOperation()->emitError("Canonicalizer failed to converge");
+      return signalPassFailure();
+    }
+  }
+  GreedyRewriteConfig config;
```

Comment:
This failed to compile last night for aarch64: https://github.com/iree-org/iree/actions/runs/11831213438/job/32966040892#step:8:6429

```
[5775/7439] Building CXX object compiler/src/iree/compiler/Codegen/Common/CMakeFiles/iree_compiler_Codegen_Common_Common.objects.dir/ConfigTrackingCanonicalizer.cpp.o
FAILED: compiler/src/iree/compiler/Codegen/Common/CMakeFiles/iree_compiler_Codegen_Common_Common.objects.dir/ConfigTrackingCanonicalizer.cpp.o 
/opt/rh/gcc-toolset-13/root/usr/bin/c++  -I/work/c -I/work/iree-build -I/work/c/third_party/llvm-project/llvm/include -I/work/iree-build/llvm-project/include -I/work/c/third_party/llvm-project/mlir/include -I/work/iree-build/llvm-project/tools/mlir/include -I/work/c/third_party/llvm-project/lld/include -I/work/iree-build/llvm-project/tools/lld/include -I/work/c/compiler/src -I/work/iree-build/compiler/src -I/work/c/runtime/src -I/work/iree-build/runtime/src -I/work/iree-build/runtime/src/iree/base/internal/flatcc -I/work/c/llvm-external-projects/iree-dialects/include -I/work/iree-build/llvm-external-projects/mlir-iree-dialects/include -I/work/c/compiler/bindings/c/. -I/work/iree-build/compiler/bindings/c/. -isystem /work/c/third_party/flatcc/include -O3 -DNDEBUG -std=gnu++17 -fPIC -fvisibility=hidden -fno-rtti -fno-exceptions -Wall -Werror -Wno-error=deprecated-declarations -Wno-address -Wno-address-of-packed-member -Wno-comment -Wno-format-zero-length -Wno-uninitialized -Wno-overloaded-virtual -Wno-invalid-offsetof -Wno-sign-compare -Wno-unused-function -Wno-unknown-pragmas -Wno-unused-but-set-variable -Wno-misleading-indentation -fmacro-prefix-map=/work/c=c -pthread -I/work/c/third_party/flatcc/include/ -I/work/c/third_party/flatcc/include/flatcc/reflection/ -MD -MT compiler/src/iree/compiler/Codegen/Common/CMakeFiles/iree_compiler_Codegen_Common_Common.objects.dir/ConfigTrackingCanonicalizer.cpp.o -MF compiler/src/iree/compiler/Codegen/Common/CMakeFiles/iree_compiler_Codegen_Common_Common.objects.dir/ConfigTrackingCanonicalizer.cpp.o.d -o compiler/src/iree/compiler/Codegen/Common/CMakeFiles/iree_compiler_Codegen_Common_Common.objects.dir/ConfigTrackingCanonicalizer.cpp.o -c /work/c/compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp
/work/c/compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp: In member function ‘virtual void mlir::iree_compiler::{anonymous}::ConfigTrackingCanonicalizerPass::runOnOperation()’:
/work/c/compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp:95:21: error: storing the address of local variable ‘listener’ in ‘*this.mlir::iree_compiler::{anonymous}::ConfigTrackingCanonicalizerPass::config.mlir::GreedyRewriteConfig::listener’ [-Werror=dangling-pointer=]
   95 |     config.listener = &listener;
      |     ~~~~~~~~~~~~~~~~^~~~~~~~~~~
/work/c/compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp:94:10: note: ‘listener’ declared here
   94 |     auto listener = ConfigTrackingListener();
      |          ^~~~~~~~
/work/c/compiler/src/iree/compiler/Codegen/Common/ConfigTrackingCanonicalizer.cpp:92:25: note: ‘this’ declared here
   92 |   void runOnOperation() override {
      |                         ^~~~~~~~
cc1plus: all warnings being treated as errors
```

Reply:
(using gcc 13.3.1, for release builds: https://github.com/iree-org/iree/actions/runs/11831213438/job/32966040892#step:8:167, the clang 14.0.0 build in another workflow succeeded: https://github.com/iree-org/iree/actions/runs/11834367067/job/32974914085#step:5:18)

Reply:
this looks to me like a false positive

Reply:
Maybe. Should we change the code a bit or disable the warning/error?